cmd/samples/pso/workflow.go

```diff
@@ -0,0 +1,139 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"go.uber.org/cadence"
+	"go.uber.org/zap"
+
+	"github.com/pborman/uuid"
+	"go.uber.org/cadence/workflow"
+)
+
+type (
+	individual struct {
+		IndividualID string
+		Variable     []float64
+		Fitness      []float64
+	}
+)
+
+// ApplicationName is the task list for this sample
+const ApplicationName = "PSO"
+
+// HostID - Use a new uuid just for demo so we can run 2 host specific activity workers on same machine.
+// In real world case, you would use a hostname or ip address as HostID.
+var HostID = ApplicationName + "_" + uuid.New()
+
+var ActivityOptions = workflow.ActivityOptions{
+	ScheduleToStartTimeout: time.Second * 5,
+	StartToCloseTimeout:    time.Minute * 10,
+	HeartbeatTimeout:       time.Second * 2, // such a short timeout to make sample fail over very fast
+	RetryPolicy: &cadence.RetryPolicy{
+		InitialInterval:          time.Second,
+		BackoffCoefficient:       2.0,
+		MaximumInterval:          time.Minute,
+		ExpirationInterval:       time.Minute * 10,
+		MaximumAttempts:          5,
+		NonRetriableErrorReasons: []string{"bad-error"},
+	},
+}
+
+var QueryResult string
```

Comment:
Cannot use a mutable global variable inside the workflow code as multiple workflow instances might be executed on the same host. Either pass it through the appropriate structures or inside the workflow.Context named variable.

Reply:
I've tried to define 
```
err = workflow.SetQueryHandler(ctx, "state", func(input []byte) (string, error) {
	return ctx.Value("QueryResult").(string), nil
})
```
and then set the values like
```
ctx = workflow.WithValue(ctx, "QueryResult", "initialized")
```
but I have problems reading the latest result, it seems is not updated. Do I have to set the BackgroundActivityContext like in the recovery sample?

Reply:
You cannot share variables across workflows. As you set handler in a parent and update the value in the child it is not going to work.

Reply:
I see. I've tried but I have not figure out how to get the RunId of the child workflow from the parent workflow, I think it's necessary with the ChildID to create a query.

Also I cannot figure out how to make a query to the child workflow in the test environment.

Reply:
ExecuteChildWorkflow returns ChildWorkflowFuture. Its GetChildWorkflowExecution method returns a Future that contains WorkflowExecution structure that contains runID. This future becomes ready as soon as the child is started.

```
		childWorkflowFuture := workflow.ExecuteChildWorkflow(ctx, child, ...)
		var childWE WorkflowExecution
		err := childWorkflowFuture.GetChildWorkflowExecution().Get(&childWE);
                 if err != nil {
                      return err
                 }
                 runID = childWE.RunID

```

Reply:
I get an error `undefined: WorkflowExecution`, where it is supposed to be defined?